csrc/scheduler/matmul_heuristic_plugin.cpp

```diff
@@ -0,0 +1,210 @@
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <ir/interface_nodes.h>
+#include <mma_type.h>
+#include <scheduler/matmul_heuristic_plugin.h>
+#include <scheduler/mma_utils.h>
+#include <utils.h>
+
+#include <dlfcn.h>
+#include <cstdint>
+#include <mutex>
+
+namespace nvfuser {
+
+namespace matmul_heuristic_plugin {
+
+namespace {
+
+std::mutex plugin_mutex;
+static class PluginInterface : NonCopyable {
+ public:
+  PluginInterface() : filepath_(getNvFuserEnv("MATMUL_HEURISTIC_PLUGIN")) {
+    if (filepath_ != nullptr) {
+      std::lock_guard<std::mutex> lock(plugin_mutex);
+      library_handle_ = dlopen(filepath_, RTLD_LAZY);
+      NVF_CHECK(
+          library_handle_ != nullptr,
+          "Error occurred when loading matmul heuristic plugin ",
+          filepath_,
+          ". Error msg: ",
+          dlerror());
+    }
+  }
+
+  bool available() const {
+    return library_handle_ != nullptr;
+  }
+
+  ~PluginInterface() {
+    if (library_handle_ != nullptr) {
+      std::lock_guard<std::mutex> lock(plugin_mutex);
+      dlclose(library_handle_);
+      library_handle_ = nullptr;
+      factory_func_ptr_ = nullptr;
+    }
+  }
+
+  KernelConfig* makeConfig() {
+    NVF_ERROR(available());
+
+    if (factory_func_ptr_ == nullptr) {
+      std::lock_guard<std::mutex> lock(plugin_mutex);
+      factory_func_ptr_ =
+          (KernelConfigFactoryPointer)dlsym(library_handle_, "makeConfig");
+      NVF_CHECK(
+          factory_func_ptr_ != nullptr,
+          "Failed to load symbol \"makeConfig\" from plugin file ",
+          filepath_,
+          ". Error message: ",
+          dlerror());
+    }
+
+    return (*factory_func_ptr_)();
+  }
+
+ private:
+  char* filepath_ = nullptr;
+  void* library_handle_ = nullptr;
+  KernelConfigFactoryPointer factory_func_ptr_ = nullptr;
+} plugin;
+
+KernelConfig* defaultConfigFactory() {
+  return plugin.makeConfig();
+}
+
+thread_local KernelConfigFactoryPointer config_factory_ptr =
+    defaultConfigFactory;
+
+void setKernelConfigFactoryPointer(KernelConfigFactoryPointer func) {
+  config_factory_ptr = func == nullptr ? defaultConfigFactory : func;
+}
+
+KernelConfig* makeConfig() {
+  NVF_ERROR(config_factory_ptr != nullptr);
+  return (*config_factory_ptr)();
+}
+
+//! Utility to standardize conversion of MmaLayout to uint8_t
+uint8_t layoutToByte(MmaLayout layout) {
+  switch (layout) {
+    case MmaLayout::NN:
+      return 0;
+    case MmaLayout::NT:
+      return 1;
+    case MmaLayout::TN:
+      return 2;
+    case MmaLayout::TT:
+      return 3;
+    default:
+      return 255;
+  }
+}
+
+} // namespace
+
+bool hasPlugin() {
+  return plugin.available();
+}
+
+bool updateMatmulParams(
+    MatmulParams& params,
+    int64_t m,
+    int64_t n,
+    int64_t k,
+    int64_t batch_size,
+    MmaLayout layout,
+    const mma_utils::RolesMap& roles_map) {
+  if (!hasPlugin()) {
+    return false;
+  }
```

Comment:
This shouldn't return false if `!hasPlugin()` but the user manually specified a function?

Reply:
Right. Updated